### ISSUE TITLE
#find_by_attribute_name returns nil and produces weird SQL statement

### ISSUE THREAD

**Issue by ghost** [globalize/globalize#423]
Calling `#find_by_attribute_name` on translated object produces following output:

```
2.2.0 :001 > Place.find_by_name 'London'

  Place Load (4.0ms)  SELECT  `places`.* FROM `places` INNER JOIN `place_translations` ON `place_translations`.`place_id` = `places`.`id` WHERE `place_translations`.`name` = '--- !ruby/object:ActiveRecord::StatementCache::Substitute {}\n' AND `place_translations`.`locale` = 'de' LIMIT 1
 => nil

2.2.0 :002 > Place.where(name: 'London').first

  Place Load (4.0ms)  SELECT  `places`.* FROM `places` INNER JOIN `place_translations` ON `place_translations`.`place_id` = `places`.`id` WHERE `place_translations`.`name` = 'London' AND `place_translations`.`locale` IN ('en', 'de')  ORDER BY `places`.`id` ASC LIMIT 1
  Place::Translation Load (0.4ms)  SELECT `place_translations`.* FROM `place_translations` WHERE `place_translations`.`place_id` IN (1)
 => #<Place id: 1, name: "London", created_at: "2013-04-04 09:48:42", updated_at: "2015-02-23 11:51:41"> 
```
- ruby 2.2.0p0 (2014-12-25 revision 49005) [x86_64-linux]
- rails 4.2.0
- globalize 5.0.1

**Comment by dasnotme** [comment]
Seems that adding `joins(:translations)` before calling `find_by_attribute_name` fixed the issue.

```
2.2.0 :001 > Place.joins(:translations).find_by_name 'London'

  Place Load (0.6ms)  SELECT  `places`.* FROM `places` INNER JOIN `place_translations` ON `place_translations`.`place_id` = `places`.`id` WHERE `place_translations`.`name` = 'London' LIMIT 1
  Place::Translation Load (0.3ms)  SELECT `place_translations`.* FROM `place_translations` WHERE `place_translations`.`place_id` = 1
 => #<Place id: 1, name: "London", created_at: "2013-04-04 09:48:42", updated_at: "2015-02-23 11:51:41"> 
```

But I wonder if it is an expected behaviour.

**Comment by parndt** [comment]
Yes, because the column doesn't exist on the other model

**Comment by shioyama** [comment]
Hmm... that SQL looks pretty fishy.

**Comment by parndt** [comment]
@shioyama oh you're right!

**Comment by dasnotme** [comment]
One more thing, `#find_by` method, which according to [apidock.com](http://apidock.com/rails/v4.1.8/ActiveRecord/FinderMethods/find_by) utilizes `#where(args).take`, produces the same behaviour:

```
2.2.0 :001 > Place.find_by(name: 'London')
  Place Load (1.4ms)  SELECT  `places`.* FROM `places` INNER JOIN `place_translations` ON `place_translations`.`place_id` = `places`.`id` WHERE `place_translations`.`name` = '--- !ruby/object:ActiveRecord::StatementCache::Substitute {}\n' AND `place_translations`.`locale` IN ('de', 'en') LIMIT 1
 => nil 
```

But `#where`, on the other hand, works correctly:

```
2.2.0 :002 > Place.where(name: 'London').take
  Place Load (1.9ms)  SELECT  `places`.* FROM `places` INNER JOIN `place_translations` ON `place_translations`.`place_id` = `places`.`id` WHERE `place_translations`.`name` = 'London' AND `place_translations`.`locale` IN ('de', 'en') LIMIT 1
  Place::Translation Load (0.3ms)  SELECT `place_translations`.* FROM `place_translations` WHERE `place_translations`.`place_id` IN (1)
=> #<Place id: 1, name: "London", created_at: "2013-04-04 09:48:42", updated_at: "2015-02-23 11:51:41"> 
```

Weirder thing happens when I add some more fields to the model and in the statements:

```
2.2.0 :042 > Place.find_by(name: 'London', longitude: 1.3)
  Place Load (0.8ms)  SELECT  `places`.* FROM `places` INNER JOIN `place_translations` ON `place_translations`.`place_id` = `places`.`id` WHERE `places`.`longitude` = 'London' AND `place_translations`.`name` = '--- !ruby/object:ActiveRecord::StatementCache::Substitute {}\n' AND `place_translations`.`locale` IN ('de', 'en') LIMIT 1
 => nil 
```

Whereas, `#where` still does its job:

```
2.2.0 :043 > Place.where(name: 'London', longitude: 1.3).take
  Place Load (0.7ms)  SELECT  `places`.* FROM `places` INNER JOIN `place_translations` ON `place_translations`.`place_id` = `places`.`id` WHERE `places`.`longitude` = 1.3 AND `place_translations`.`name` = 'London' AND `place_translations`.`locale` IN ('de', 'en') LIMIT 1
  Place::Translation Load (0.3ms)  SELECT `place_translations`.* FROM `place_translations` WHERE `place_translations`.`place_id` IN (1)
=> #<Place id: 1, name: "London", longitude: 1.3, created_at: "2013-04-04 09:48:42", updated_at: "2015-02-23 11:51:41"> 
```

**Comment by shioyama** [comment]
This is probably some conflict with the tricks we do in [query methods](https://github.com/globalize/globalize/blob/f1d7b06d53a5820a9e6421ae6f4357906d98292b/lib/globalize/active_record/query_methods.rb) to make translated attributes act like normal attributes.

**Comment by dasnotme** [comment]
Yeah, it seems to be happening there https://github.com/rails/rails/blob/4-2-stable/activerecord/lib/active_record/core.rb#L179 from what I understand.

**Comment by shioyama** [comment]
Hmm... maybe [this commit](https://github.com/rails/rails/commit/77b18d7ef47f4dd6bf39853ac83cc19a3593406f) is what did it. Anyway I'll have a look at it asap, if anyone else wants to take a stab at it go ahead.

Those query method hacks are handy but couple globalize pretty tightly to activerecord... there was a request earlier to make them optional (something like `Post.friendly.find(...)` in friendly_id), maybe that's an idea we could consider.

**Comment by colepanike** [comment]
I'm seeing this too. I'm using the rails42 branch to deal with this: https://github.com/rails/rails/issues/18739 Have you got a fix?

**Comment by kiliancs** [comment]
This is still a problem with rails 4.2.1 and globalize 5.0.1.
What version combination should I use to avoid this problem for now?

**Comment by krisdigital** [comment]
The only fix that worked for me is using where().take instead of find_by. Using find_by, the globalize parameters are strapped in the AR query_methods create_binds method, because of

     @klass.columns_hash.include?(column.to_s)
 
and I could not find a way to fix this without introducing new issues.

**Comment by caesarsol** [comment]
Hi, any news on this? 

Is https://github.com/globalize/globalize/issues/449 working? 

thanks :)

**Comment by shioyama** [comment]
Trying to reproduce this with mysql and I don't see the issue. Is this postgres-specific? I'm testing with mysql + sqlite.

**Comment by shioyama** [comment]
Don't think this is resolved, re-opening.

**Comment by shioyama** [comment]
I was only able to reproduce this if the column name (e.g. `name` above) exists **both** on the table and on its translation table. If this is the case, then you get very strange resutls. Globalize should probably handle this situation gracefully, but it doesn't, and you should not have the same column on both tables (we should add a note in the readme about this).

Can you confirm that this is the situation (i.e. that `name` is on both tables?) This was the case in #458 which said was the same as this one.

**Comment by tomash** [comment]
Is this still a thing?

**Comment by parndt** [comment]
nah, this was addressed by #557
